File: coting-test-by-kotlin-gradle/src/main/kotlin/gradle/cotingtest/console/codility/Demo11st.java
package gradle.cotingtest.console.codility;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toList;

public class Demo11st {

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println("# = " + solution.solution(new int[]{1, 2, 3}));
        System.out.println("# = " + solution.solution(new int[]{-1, -3}));
        System.out.println("# = " + solution.solution(new int[]{1, 3, 6, 4, 1, 2}));
    }

    static class Solution {
        public int solution(int[] A) {
            List<Integer> positiveNums = Arrays.stream(A)
                    .filter(n -> n > 0)
                    .boxed()
                    .sorted(Comparator.reverseOrder())
                    .collect(toList());

            if (positiveNums.isEmpty()) {
                return 1;
            }
            int max = positiveNums.get(0);
            List<Integer> removedNums = removedNums(max, positiveNums);

            if (removedNums.isEmpty()) {
                return max + 1;
            }

            return removedNums.get(0);
        }

        private List<Integer> removedNums(int max, List<Integer> positiveNums) {
            List<Integer> removedNums = IntStream.range(1, max + 1)
                    .boxed()
                    .sorted(Comparator.reverseOrder())
                    .collect(toList());

            for (Integer num : positiveNums) {
                removedNums.remove(num);
            }

            return removedNums;
        }
    }
}

File: coding-test-by-jvm/src/codingtest/codility/Demo11st.kt
package codingtest.codility

import java.util.Comparator
import java.util.stream.Collectors
import java.util.stream.Collectors.toList
import java.util.stream.IntStream
import kotlin.streams.toList

class Solution {

    fun solution(A: IntArray): Int {
        val positiveNums: List<Int> = A.toList().stream()
                .filter { it > 0 }
                .sorted(kotlin.Comparator.reverseOrder())
                .collect(toList())
        if (positiveNums.isEmpty()) {
            return 1
        }
        val max = positiveNums[0]
        val removedNums = removeNums(max, positiveNums)

        if (removedNums.isEmpty()) {
            return max + 1
        }

        return removedNums[0]
    }

    private fun removeNums(max: Int, positiveNums: List<Int>): List<Int> {
        var removedNums = IntStream.range(1, max + 1)
                .boxed()
                .sorted(Comparator.reverseOrder())
                .collect(toList())

        for (num in positiveNums) {
            removedNums.remove(num)
        }

        return removedNums.toList()
    }
}

fun main() {
    println(Solution().solution(intArrayOf(1, 2, 3))) // 4
    println(Solution().solution(intArrayOf(-1, -3))) // 1
    println(Solution().solution(intArrayOf(1, 3, 6, 4, 1, 2))) // 5
}
